contracts/EscrowedNori.sol

```diff
@@ -0,0 +1,725 @@
+import "@openzeppelin/contracts-upgradeable/utils/math/MathUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC777/IERC777RecipientUpgradeable.sol";
+import "./ERC777PresetPausablePermissioned.sol";
+import "./BridgedPolygonNORI.sol";
+import "./Removal.sol";
+import {RemovalUtils} from "./RemovalUtils.sol";
+
+/*
+Open Questions:
+- do we need to set up an escrow schedule with the full amount associated with the agreement?
+  or is it ok that the amount of the escrow schedule actually changes over time either with the addition
+  of new NORI or with the revocation of some?
+  - seems to me that the amount changing gradually makes more sense, because if a supplier only sells one of their
+  removals we actually want that single amount to release gradually over 10 years so that we have recourse to reclaim
+  those insurance tokens on behalf of that removal at some point in the future
+
+- do we need to allow specification of end time? or perhaps of duration? or can we assume always 10 years
+  and use that hardcoded constant?
+
+  - do we need any ability to create escrow schedules before tokens are received on behalf of a removal?
+      I would argue that we don't, and that the only placed escrow schedules should be created is in tokensReceived > depositFor
+
+  - should we set this up so that if a token is deposited on behalf of a removal id that DOESN'T have an entry for its schedule
+  -   startTime that we use the year (in epoch seconds) of the removal's vintage as the startime? it would allow us to migrate
+  -   seamlessly to the far superior world where the escrow schedules aren't so arbitrary and actually represent exactly
+  -   10 years for each removal.
+
+  - When it comes to keeping track of tokens that have been revoked, is it sufficient to keep a running total for any given escrow schedule,
+  and then to rely on events for more granular examination at the level of removalid and to see the times at which things were revoked?
+  - Do we need any additional information captured when a revocation happens?  Do we want to allow a default mechanism for recapturing ALL available tokens?
+  I believe this is currently happening by specifying an amount of 0
+ */
+
+/*
+TODO LIST:
+- handle escrow schedule transferability (both batch and single) that transfers full grants.
+  -  pretty sure we can't do this in any other way because it's too unclear which escrow schedules the transferred tokens belong to
+
+- do we need a mechanism of removing escrow schedule ids from a supplier's list once the escrow schedule has been completely
+- vested AND claimed? so that the process of withdrawing tokens is more gas efficient and iterates less??
+  - old schedules could still be kept around and indexed by a different collection, for view purposes.
+
+  - what are all the potential view functions we need here?
+
+  - update all the natspec comments
+
+  - tests tests tests!
+ */
+
+uint256 constant SECONDS_IN_TEN_YEARS = 31_536_000;
```

Comment:
What about leap years though?

[FYI solidity also has builtin globals for time (excluding years for the above reason)](https://docs.soliditylang.org/en/latest/units-and-global-variables.html#time-units)

Reply:
mmm yeah... I was already on the fence about hard-coding something like this, since we maybe want to leave open the possibility of custom insurance terms or changes to that policy in the future.  regardless, this leap year concern alone is probably enough to preclude us from hard-coding 10 years (though I would argue that 10 years +/- 1 day for the unaccountable potential leap year doesn't matter at all for our purposes here since there is already so much that's arbitrary about these insurance tokens, such as the percentage we're holding back in the first place)

Reply:
use 365.2422 days / year to calculate our 10-year value in seconds??